### PR TITLE
aggressively reclaim disk space after data have been deleted

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Periodically get tombstones number from sst properties for a range, if there are too many tombstones, trigger manual compaction for this range. 
In time series situation, only keep latest several days or months' data, the old data would be deleted periodically. The amount of data is very huge in this situation, disk space is a sensitive resource.

This pr:
1) periodically get tombstones number from sst properties, and trigger manual compaction when needed.
2) involve manual compaction once when multiple continuous regions need compacting.

### CODE REVIEW

**Comment by ngaut** [comment]
Could you add description about the actual results after this patch?

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -534,6 +544,11 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         self.apply_res_receiver = Some(rx);
         box_try!(self.apply_worker.start(apply_runner));
 
+        let (tx, rx) = mpsc::channel();
+        let space_check_runner = SpaceCheckRunner::new(Arc::clone(&self.kv_engine), tx);
```
Comment (huachaohuang) [inline src/raftstore/store/store.rs]: Can we merge this worker with the `compact_worker`? It seems overwhelmed to create a worker for each task.
BTW, can you use the `sendch` in the raft store to send back the result?

**Comment by zhangjinpeng1987** [comment]
@huachaohuang I have a document that record the test result.

File: src/raftstore/store/store.rs (hunk 1 above)
Comment (zhangjinpeng1987) [reply src/raftstore/store/store.rs]: done

File: src/raftstore/store/worker/compact.rs [hunk 2]
```diff
@@ -11,32 +11,62 @@
...
+pub enum Task {
+    Compact {
+        cf_name: String,
+        start_key: Option<Key>, // None means smallest key
+        end_key: Option<Key>,   // None means largest key
+    },
+
+    Check {
+        ranges: BTreeSet<Key>,
+        min_num_del: u64,
```
Comment (huachaohuang) [inline src/raftstore/store/worker/compact.rs]: `min_num_del` looks weird, can we use a better name?

File: src/raftstore/store/worker/compact.rs [hunk 3]
```diff
@@ -11,32 +11,62 @@
...
+                cf_name,
+                start_key.as_ref().map(|k| escape(k)),
+                end_key.as_ref().map(|k| escape(k))
+            ),
+            Task::Check {
+                ref ranges,
+                tombstones_threshold,
+            } => write!(
+                f,
+                "Space check ranges count {}, min num del {}",
```
Comment (huachaohuang) [inline src/raftstore/store/worker/compact.rs]: s/min num del/tombstones threshold/

**Comment by BusyJay** [comment]
Any updates?

**Comment by huachaohuang** [comment]
LGTM

File: src/raftstore/store/msg.rs [hunk 4]
```diff
@@ -166,6 +167,11 @@ pub enum Msg {
 
     // Compaction finished event
     CompactedEvent(CompactedEvent),
+
+    // Space check result
+    SpaceCheckResult {
```
Comment (BusyJay) [inline src/raftstore/store/msg.rs]: Why not do it in the compact worker immediately?

File: src/raftstore/store/store.rs [hunk 5]
```diff
@@ -151,6 +151,8 @@ pub struct Store<T, C: 'static> {
     pub apply_worker: Worker<ApplyTask>,
     apply_res_receiver: Option<StdReceiver<ApplyTaskRes>>,
 
+    last_checked_key: Option<Key>,
```
Comment (huachaohuang) [inline src/raftstore/store/store.rs]: Using `last_compact_checked_key` is more clear.

File: src/raftstore/store/worker/compact.rs [hunk 6]
```diff
@@ -149,4 +318,83 @@ mod test {
...
+            mvcc_put(&engine, k.as_bytes(), v.as_bytes(), 1, 2);
+        }
+        engine.flush_cf(cf, true).unwrap();
+
+        let (s, e) = (data_key(b"k5"), data_key(b"k9"));
+        let (entries, version) = get_range_entries_and_versions(&engine, cf, &s, &e).unwrap();
+        assert_eq!(entries, 5);
+        assert_eq!(version, 5);
+
+        let mut ranges_to_check = BTreeSet::new();
```
Comment (huachaohuang) [inline src/raftstore/store/worker/compact.rs]: Add more ranges to cover the neighbor ranges case?

Comment (zhangjinpeng1987) [reply src/raftstore/store/worker/compact.rs]: done

**Comment by huachaohuang** [comment]
LGTM

File: src/raftstore/store/store.rs [hunk 7]
```diff
@@ -1876,24 +1879,46 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-            for &cf in &[CF_DEFAULT, CF_WRITE] {
-                let task = CompactTask {
-                    cf_name: String::from(cf),
-                    start_key: Some(keys::enc_start_key(peer.region())),
-                    end_key: Some(keys::enc_end_key(peer.region())),
-                };
-                if let Err(e) = self.compact_worker.schedule(task) {
-                    error!("{} failed to schedule compact task: {}", self.tag, e);
+        if self.compact_worker.is_busy() {
+            info!("compact worker is busy, check space redundancy next time");
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: You can return early to reduce indent.

File: src/raftstore/store/store.rs [hunk 8]
```diff
@@ -151,6 +151,8 @@ pub struct Store<T, C: 'static> {
     pub apply_worker: Worker<ApplyTask>,
     apply_res_receiver: Option<StdReceiver<ApplyTaskRes>>,
 
+    last_compact_checked_key: Option<Key>,
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Why not just use `Key` and initialize it with `DATA_MIN_KEY`?

File: src/raftstore/store/store.rs [hunk 9]
```diff
@@ -1876,24 +1879,46 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+                Some(key) => key,
+                None => {
+                    ranges_need_check.insert(DATA_MIN_KEY.to_vec());
+                    DATA_MIN_KEY.to_vec()
                 }
+            };
+
+            // Collect multiple continuous ranges.
+            let left_ranges = self.region_ranges
+                .range((Included(last_compact_checked_key), Unbounded::<Key>));
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Why `Included`?

File: src/raftstore/store/store.rs [hunk 10]
```diff
@@ -1876,24 +1879,46 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            for (count, (key, _)) in left_ranges.enumerate() {
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: `left_ranges.take(self.cfg.region_compact_check_step)`.

File: src/raftstore/store/store.rs [hunk 11]
```diff
@@ -1876,24 +1879,46 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+                ranges_need_check.insert(key.to_vec());
+                if count > self.cfg.region_compact_check_step as usize {
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: I think it should be `>=`?

File: src/raftstore/store/worker/compact.rs [hunk 12]
```diff
@@ -11,32 +11,63 @@
...
+pub enum Task {
+    Compact {
+        cf_name: String,
+        start_key: Option<Key>, // None means smallest key
+        end_key: Option<Key>,   // None means largest key
+    },
+
+    CheckAndCompact {
+        cf_names: Vec<String>,     // Column families need to compact
+        ranges: BTreeSet<Key>,     // Ranges need to check
```
Comment (BusyJay) [inline src/raftstore/store/worker/compact.rs]: Why not use `Vec<Key>`?

File: src/raftstore/store/store.rs (hunk 8 above)
Comment (zhangjinpeng1987) [reply src/raftstore/store/store.rs]: We still need handle None after we have finished a round.

File: src/raftstore/store/store.rs (hunk 7 above)
Comment (zhangjinpeng1987) [reply src/raftstore/store/store.rs]: After refactor, there are 3 branches in this function, we must register tick 3 times if return early to reduce indent.

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL

**Comment by BusyJay** [comment]
LGTM

File: src/raftstore/store/store.rs [hunk 13]
```diff
@@ -1876,24 +1879,52 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+                Unbounded::<Key>,
+            ));
+            ranges_need_check.extend(
+                left_ranges
+                    .take(self.cfg.region_compact_check_step as usize)
+                    .map(|(k, _)| k.to_owned()),
+            );
+
+            // Update last_compact_checked_key.
+            let largest_key = self.region_ranges.keys().next_back().unwrap().to_vec();
```
Comment (huachaohuang) [inline src/raftstore/store/store.rs]: `self.region_ranges.keys().last()`?

File: src/raftstore/store/store.rs [hunk 14]
```diff
@@ -1876,24 +1879,52 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            let last_key = ranges_need_check[ranges_need_check.len() - 1].clone();
```
Comment (huachaohuang) [inline src/raftstore/store/store.rs]: `ranges_need_check.last()`?

**Pull request merged**